### PR TITLE
Fuzz Fuzz Fuzz

### PR DESCRIPTION
**Pull request by Soveu**
I'm not done yet, expect more fuzzers tommorow (for `ArrayVecDrain` for example)

### CODE REVIEW

**Comment by Soveu** [comment]
Previous `TinyVecDrain` did not implement `nth`, `next_back` and `size_hint`, so I used it as a reason to re-implement it to this
```rust
pub enum TinyVecDrain<'p, A: Array> {
  Inline(ArrayVecDrain<'p, A::Item>),
  Heap(vec::Drain<'p, A::Item>),
}
```

**Comment by Soveu** [comment]
While running `cargo hfuzz arrayvec_iter` the fuzzer crashes
```
The application panicked (crashed).
  Box<Any>
in fuzz/src/bin/arrayvec_iter.rs, line 10
```
Line 10 is where the `arbitrary_stateful_operations` macro begins
@Shnatsel @jakubadamw

**Comment by Shnatsel** [comment]
Is that the first panic you're seeing? I've seen that as the final panic _after_ the one that actually caused the crash.

**Comment by Soveu** [comment]
With `run-debug` the test panics once, continuing execution just makes the program exit

**Comment by Soveu** [comment]
```
soveu@localhost ~/g/t/fuzz (fuzzfuzzfuzz)> 
cargo hfuzz run-debug arrayvec_iter hfuzz_workspace/arrayvec_iter/*.fuzz
/usr/bin/ld.gold
warning: unused import: `convert::TryInto`
 --> fuzz/src/bin/arrayvec_iter.rs:3:11
  |
3 | use std::{convert::TryInto, fmt::Debug};
  |           ^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: unused import: `tinyvec_fuzz::ArbRange`
 --> fuzz/src/bin/arrayvec_iter.rs:6:5
  |
6 | use tinyvec_fuzz::ArbRange;
  |     ^^^^^^^^^^^^^^^^^^^^^^

warning: 2 warnings emitted

    Finished dev [unoptimized + debuginfo] target(s) in 0.03s
GNU gdb (GDB) Fedora 9.1-5.fc32
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from hfuzz_target/x86_64-unknown-linux-gnu/debug/arrayvec_iter...
warning: Missing auto-load script at offset 0 in section .debug_gdb_scripts
of file /home/soveu/git/tinyvec/fuzz/hfuzz_target/x86_64-unknown-linux-gnu/debug/arrayvec_iter.
Use `info auto-load python-scripts [REGEXP]' to list them.
Breakpoint 1 at 0x1339b3: file src/libstd/panicking.rs, line 575.
Starting program: /home/soveu/git/tinyvec/fuzz/hfuzz_target/x86_64-unknown-linux-gnu/debug/arrayvec_iter 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Backtrace (most recent call first):
  File "/home/soveu/git/tinyvec/fuzz/src/bin/arrayvec_iter.rs", line 10, in arrayvec_iter::op::Op::execute_and_compare
    arbitrary_stateful_operations! {
  File "/home/soveu/git/tinyvec/fuzz/src/bin/arrayvec_iter.rs", line 65, in arrayvec_iter::fuzz_cycle
    op.execute_and_compare(&mut model, &mut tested);
  File "/home/soveu/git/tinyvec/fuzz/src/bin/arrayvec_iter.rs", line 76, in arrayvec_iter::main::{{closure}}
    let _ = fuzz_cycle(data);
  File "/home/soveu/.cargo/registry/src/github.com-1ecc6299db9ec823/honggfuzz-0.5.49/src/lib.rs", line 329, in honggfuzz::fuzz
    closure(&mmap);
  File "/home/soveu/git/tinyvec/fuzz/src/bin/arrayvec_iter.rs", line 75, in arrayvec_iter::main
    fuzz!(|data: &[u8]| {

The application panicked (crashed).
  Box<Any>
in fuzz/src/bin/arrayvec_iter.rs, line 10
thread: main

Breakpoint 1, std::panicking::rust_panic () at src/libstd/panicking.rs:575
575	src/libstd/panicking.rs: No such file or directory.
#0  std::panicking::rust_panic () at src/libstd/panicking.rs:575
#1  0x000055555568796a in std::panicking::rust_panic_with_hook () at src/libstd/panicking.rs:545
#2  0x000055555558072d in std::panicking::begin_panic (msg=...)
    at /home/soveu/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panicking.rs:456
#3  0x000055555557f78f in arrayvec_iter::op::Op::execute_and_compare (self=..., 
    model=0x7fffffffd9b0, tested=0x7fffffffd8c0) at fuzz/src/bin/arrayvec_iter.rs:10
#4  0x000055555557adf4 in arrayvec_iter::fuzz_cycle (data=...) at fuzz/src/bin/arrayvec_iter.rs:65
Missing separate debuginfos, use: dnf debuginfo-install glibc-2.31-4.fc32.x86_64 libgcc-10.1.1-1.fc32.x86_64
--Type <RET> for more, q to quit, c to continue without paging--c
#5  0x000055555557ffc3 in arrayvec_iter::main::{{closure}} (data=...) at fuzz/src/bin/arrayvec_iter.rs:76
#6  0x000055555557a2cf in honggfuzz::fuzz (closure=...) at /home/soveu/.cargo/registry/src/github.com-1ecc6299db9ec823/honggfuzz-0.5.49/src/lib.rs:329
#7  0x000055555557ae7e in arrayvec_iter::main () at fuzz/src/bin/arrayvec_iter.rs:75
(gdb) c
Continuing.
[Inferior 1 (process 7850) exited with code 0145]
(gdb) 
```

**Comment by Soveu** [comment]
Could it be because of [this](https://github.com/jakubadamw/rutenspitz/blob/97d27d2a36fdc76d6572b1b271fdccced82f3f15/lib/src/lib.rs#L15)

**Comment by Soveu** [comment]
Ok, I need to manually set a hook
```rust
fn main() -> Result<(), ()> {
  //better_panic::install();
  std::panic::set_hook(Box::new(|panic_info| {
    if let Some(outpanic) = panic_info.payload().downcast_ref::<rutenspitz::OutcomePanic>() {
        eprintln!("{}", outpanic.0);
        std::process::abort();
    }
  }));


  loop {
    fuzz!(|data: &[u8]| {
      let _ = fuzz_cycle(data);
    });
  }
}
```

**Comment by Soveu** [comment]
ugh, the message could at least say which outcome comes from model
`The return values aren't equal: 'None' != 'Some(267443951)'`

**Comment by Soveu** [comment]
Okay, another bug in `ArrayVecIterator::nth_back`

File: src/arrayvec.rs [hunk 1]
```diff
@@ -1171,15 +1171,20 @@ impl<A: Array> DoubleEndedIterator for ArrayVecIterator<A> {
...
-    /* n is in [0..self.tail - self.base] range, so in u16 range */
-    self.tail = self.base + n as u16;
+    let base = self.base as usize;
+    let tail = self.tail as usize;
+    let slice = &mut self.data.as_slice_mut()[base..tail];
+    let n = n.saturating_add(1);
+
+    if let Some(n) = slice.len().checked_sub(n) {
+      let item = &mut slice[n];
+      /* n is in [0..self.tail - self.base] range, so in u16 range */
```
Comment (Lokathor) [inline src/arrayvec.rs:1181]: In the future, things like this are a good chance to use `debug_assert`, though we don't need another commit to change that here.

**Pull request merged**